README.md
# sdx-sequence

[![Build Status](https://travis-ci.org/ONSdigital/sdx-sequence.svg?branch=master)](https://travis-ci.org/ONSdigital/sdx-sequence) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/6e6856e9c191481ebeed8c10b70cfc16)](https://www.codacy.com/app/ons-sdc/sdx-sequence?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ONSdigital/sdx-sequence&amp;utm_campaign=Badge_Grade) [![codecov](https://codecov.io/gh/ONSdigital/sdx-sequence/branch/master/graph/badge.svg)](https://codecov.io/gh/ONSdigital/sdx-sequence)

Scalable service for generating sequences for SDX.

## Prerequisites

A running instance of Postgres DB. The service connects to `postgres://sdx@sdxlocalhost:5432/sdx` by default.


## Installation
This application presently installs required packages from requirements files:
- `requirements.txt`: packages for the application, with hashes for all packages: see https://pypi.org/project/hashin/
- `test-requirements.txt`: packages for testing and linting

It's also best to use `pyenv` and `pyenv-virtualenv`, to build in a virtual environment with the currently recommended version of Python.  To install these, see:
- https://github.com/pyenv/pyenv
- https://github.com/pyenv/pyenv-virtualenv
- (Note that the homebrew version of `pyenv` is easiest to install, but can lag behind the latest release of Python.)

### Getting started
Once your virtual environment is set, install the requirements:
```shell
$ make build
```

To test, first run `make build` as above, then run:
```shell
$ make test
```

### Docker

```bash
$ docker build -t sdx-sequence
```

## Usage

Start the sdx-sequence service using the following command:

```shell
$ python server.py
```

## API

There are three endpoints for the three types of sequences:
 * `GET /sequence`
 * `GET /batch-sequence`
 * `GET /image-sequence`

There is also a health check endpoint, which returns a json response with key/value pairs describing the service state:
 * `GET /healthcheck`

## Example

Curl a request to any one of the endpoints:
```
curl http://localhost:5000/sequence
```

The response should look something like:
```
{
    sequence_no: 1000
}
```

## Configuration

The following envioronment variables can be set for non cloudfoundry use:

| Environment Variable           | Default                               | Description
|--------------------------------|---------------------------------------|----------------
| SDX_SEQUENCE_POSTGRES_HOST     | `127.0.0.1`                           | The PostgreSQL host
| SDX_SEQUENCE_POSTGRES_PORT     | `5432`                                | The PostgreSQL port
| SDX_SEQUENCE_POSTGRES_NAME     | `sdx`                                 | The PostgreSQL database
| SDX_SEQUENCE_POSTGRES_USER     | `sdx`                                 | The PostgreSQL user
| SDX_SEQUENCE_POSTGRES_PASSWORD | `sdx`                                 | The PostgreSQL password

For cloudfoundry only the DB_URL needs to be set in vcap services the format:

postgres://DB_USER:DB_PASSWORD@DB_HOST:DB_PORT/DB_NAME

### License

Copyright ©‎ 2016, Office for National Statistics (https://www.ons.gov.uk)

Released under MIT license, see [LICENSE](LICENSE) for details.


CHANGELOG.md
### Unreleased
  - Add openapi.yml
  
### 2.6.4 2020-06-09
  - Remove Cloudfoundry deployment files
  
### 2.6.3 2020-05-21  
  - Updated packages

### 2.6.2 2019-11-08
  - Add python 3.8 to travis builds

### 2.6.1 2019-09-26
  - Update packages, including Werkzeug that had a security vulnerability

### 2.6.0 2019-05-31
  - Remove python 3.4 and 3.5 from travis builds
  - Add python 3.7 to travis builds
  - Upgrade urllib3, Jinja2 and sqlalchemy to fix security issues.
  - Upgrade various packages.

### 2.5.1 2019-02-19
  - Update packages to fix security issues
  - Fix logging to output in a standard way

### 2.5.0 2018-01-04
  - Add /info healthcheck endpoint

### 2.4.0 2017-11-21
  - Add Cloudfoundry deployment files
  - Remove use of sdx-common for logging

### 2.3.0 2017-11-01
  - Begin using pytest as default test runner
  - Add all service config to config file

### 2.2.0 2017-09-25
  - Removed SDX common clone in docker
  - Remove JSON logging
  - Allow batches of sequence numbers to be requested

### 2.1.0 2017-07-25
  - Change all instances of ADD to COPY in Dockerfile
  - Remove use of SDX_HOME variable in makefile
  - Fix database tests on travis

### 2.0.0 2017-07-10
  - Use Postgres backend via SQLAlchemy ORM.
  - Add environment variable details to README
  - Correct license attribution
  - Add codacy badge
  - Adding sdx-common functionality
  - Updating logger format using sdx-common
  - Fix Dockerfile to load dependencies before using them.
  - Add support for codecov to see unit test coverage
  - Update and pin version of sdx-common to 0.7.0

### 1.3.1 2017-03-15
  - Log version number on startup
  - Add change log

### 1.3.0 2016-12-13
  - Add new `/json-sequence` endpoint to support census processing

### 1.2.0 2016-10-24
  - Add new `/healthcheck` endpoint

### 1.1.0 2016-09-19
  - Add configurable log level

### 1.0.0 2016-07-13
  - Initial release


tests/__init__.py
from sqlalchemy import create_engine
import testing.postgresql

import settings
from sequences import create_sequences

# Launch new PostgreSQL server
Postgres = testing.postgresql.PostgresqlFactory(cache_initialized_db=False)
postgresql = Postgres()
settings.DB_URL = postgresql.url()
engine = create_engine(postgresql.url())
create_sequences(engine)


tests/test_server.py
'''
Flask functional tests
'''

import unittest
import json
import testing.postgresql
import server


@testing.postgresql.skipIfNotInstalled
class SequenceNoTestCase(unittest.TestCase):

    def setUp(self):
        server.app.config['TESTING'] = True
        self.app = server.app.test_client()

    def test_get_sequence(self):
        sequence_resp = self.app.get('/sequence')
        sequence_json = json.loads(sequence_resp.get_data(as_text=True))
        self.assertEqual(200, sequence_resp.status_code)
        sequence_start = 1000
        sequence_range = 9999
        sequence_no = sequence_json['sequence_no']
        self.assertTrue(sequence_no >= sequence_start, "Sequence should be greater than 1000 was {}".format(sequence_no))
        self.assertTrue(sequence_no <= sequence_range, "Sequence should be less than 9999 was {}".format(sequence_no))

    def test_json_sequence(self):
        sequence_resp = self.app.get('/json-sequence')
        sequence_json = json.loads(sequence_resp.get_data(as_text=True))
        self.assertEqual(200, sequence_resp.status_code)
        sequence_start = 1
        sequence_range = 999999999
        sequence_no = sequence_json['sequence_no']
        self.assertTrue(sequence_no >= sequence_start, "Sequence should be greater than 1 was {}".format(sequence_no))
        self.assertTrue(sequence_no <= sequence_range, "Sequence should be less than 999999999 was {}".format(sequence_no))

    def test_batch_sequence(self):
        sequence_resp = self.app.get('/batch-sequence')
        sequence_json = json.loads(sequence_resp.get_data(as_text=True))
        self.assertEqual(200, sequence_resp.status_code)
        sequence_start = 30000
        sequence_range = 39999
        sequence_no = sequence_json['sequence_no']
        self.assertTrue(sequence_no >= sequence_start, "Sequence should be greater than 30000 was {}".format(sequence_no))
        self.assertTrue(sequence_no <= sequence_range, "Sequence should be less than 39999 was {}".format(sequence_no))

    def test_image_sequence(self):
        sequence_resp = self.app.get('/image-sequence')
        sequence_json = json.loads(sequence_resp.get_data(as_text=True))
        self.assertEqual(200, sequence_resp.status_code)
        sequence_start = 1
        sequence_range = 999999999
        sequence_no = sequence_json['sequence_no']
        self.assertTrue(sequence_no >= sequence_start, "Sequence should be greater than 1 was {}".format(sequence_no))
        self.assertTrue(sequence_no <= sequence_range, "Sequence should be less than 999999999 was {}".format(sequence_no))

    def tearDown(self):
        pass


@testing.postgresql.skipIfNotInstalled
class SequenceListTestCase(unittest.TestCase):

    def setUp(self):
        server.app.config['TESTING'] = True
        self.app = server.app.test_client()

    def test_get_sequence_bad_query(self):
        sequence_resp = self.app.get('/sequence?n=abcd')
        self.assertEqual(400, sequence_resp.status_code)

    def test_get_sequence(self):
        seq_min = 1000
        seq_max = 9999

        for n in range(0, 13):
            with self.subTest(n=n):
                sequence_resp = self.app.get('/sequence?n={0}'.format(n))
                sequence_json = json.loads(sequence_resp.get_data(as_text=True))
                self.assertEqual(200, sequence_resp.status_code)
                sequence_list = sequence_json['sequence_list']
                self.assertEqual(n, len(sequence_list))
                self.assertEqual(n, len(set(sequence_list)))
                self.assertTrue(all(seq_min <= i <= seq_max) for i in sequence_list)

    def test_json_sequence(self):
        seq_min = 1
        seq_max = 999999999

        for n in range(0, 13):
            with self.subTest(n=n):
                sequence_resp = self.app.get('/json-sequence?n={0}'.format(n))
                sequence_json = json.loads(sequence_resp.get_data(as_text=True))
                self.assertEqual(200, sequence_resp.status_code)
                sequence_list = sequence_json['sequence_list']
                self.assertEqual(n, len(sequence_list))
                self.assertEqual(n, len(set(sequence_list)))
                self.assertTrue(all(seq_min <= i <= seq_max) for i in sequence_list)

    def test_batch_sequence(self):
        seq_min = 30000
        seq_max = 39999

        for n in range(0, 13):
            with self.subTest(n=n):
                sequence_resp = self.app.get('/batch-sequence?n={0}'.format(n))
                sequence_json = json.loads(sequence_resp.get_data(as_text=True))
                self.assertEqual(200, sequence_resp.status_code)
                sequence_list = sequence_json['sequence_list']
                self.assertEqual(n, len(sequence_list))
                self.assertEqual(n, len(set(sequence_list)))
                self.assertTrue(all(seq_min <= i <= seq_max) for i in sequence_list)

    def test_image_sequence(self):
        seq_min = 1
        seq_max = 999999999

        for n in range(0, 13):
            with self.subTest(n=n):
                sequence_resp = self.app.get('/image-sequence?n={0}'.format(n))
                sequence_json = json.loads(sequence_resp.get_data(as_text=True))
                self.assertEqual(200, sequence_resp.status_code)
                sequence_list = sequence_json['sequence_list']
                self.assertEqual(n, len(sequence_list))
                self.assertEqual(n, len(set(sequence_list)))
                self.assertTrue(all(seq_min <= i <= seq_max) for i in sequence_list)

    def tearDown(self):
        pass


if __name__ == '__main__':
    unittest.main()


tests/test_sequence.py
from server import app
import unittest
import json
import mock


class TestSequenceService(unittest.TestCase):

    sequence_endpoint = "/sequence"
    batch_sequence_endpoint = "/batch-sequence"
    image_sequence_endpoint = "/image-sequence"
    json_sequence_endpoint = "/json-sequence"

    def setUp(self):
        # creates a test client
        self.app = app.test_client()
        # propagate the exceptions to the test client
        self.app.testing = True

    def mock_sequence_response(self, endpoint, mock_value, seq_start, seq_range, expected_sequence_no=False):
        # Use the loop index as to mock a return param
        with mock.patch('server.sequence_values', return_value=iter([mock_value])):
            r = self.app.get(endpoint)

            expected = expected_sequence_no if expected_sequence_no else seq_start + (mock_value - 1) % seq_range

            actual_response = json.loads(r.data.decode('UTF8'))
            expected_response = {'sequence_no': expected, 'sequence_list': [expected]}

            self.assertEqual(actual_response, expected_response)

    def test_increments_sequence(self):
        sequence_start = 1000
        sequence_range = 9000

        test_start = 1
        test_end = 10

        for i in range(test_start, test_end):
            self.mock_sequence_response(self.sequence_endpoint, i, sequence_start, sequence_range)

    def test_increment_wraps_sequence(self):
        sequence_start = 1000
        sequence_range = 9000

        test_start = 8998
        test_end = 9003

        for i in range(test_start, test_end):
            self.mock_sequence_response(self.sequence_endpoint, i, sequence_start, sequence_range)

    def test_increments_batch_sequence(self):
        sequence_start = 30000
        sequence_range = 10000

        test_start = 8998
        test_end = 9003

        for i in range(test_start, test_end):
            self.mock_sequence_response(self.batch_sequence_endpoint, i, sequence_start, sequence_range)

    def test_increment_wraps_batch_sequence(self):
        sequence_start = 30000
        sequence_range = 10000

        test_start = 39998
        test_end = 40003

        for i in range(test_start, test_end):
            self.mock_sequence_response(self.batch_sequence_endpoint, i, sequence_start, sequence_range)

    def test_increments_image_sequence(self):
        sequence_start = 1
        sequence_range = 1000000000

        test_start = 1
        test_end = 10

        for i in range(test_start, test_end):
            expected_seq_no = i % sequence_range
            self.mock_sequence_response(self.image_sequence_endpoint, i, sequence_start, sequence_range, expected_sequence_no=expected_seq_no)

    def test_increment_wraps_image_sequence(self):
        sequence_start = 1
        sequence_range = 1000000000

        test_start = 9999999998
        test_end = 1000000005

        for i in range(test_start, test_end):
            expected_seq_no = i % sequence_range
            self.mock_sequence_response(self.image_sequence_endpoint, i, sequence_start, sequence_range, expected_sequence_no=expected_seq_no)

    def test_increments_json_sequence(self):
        sequence_start = 1
        sequence_range = 1000000000

        test_start = 1
        test_end = 10

        for i in range(test_start, test_end):
            expected_seq_no = i % sequence_range
            self.mock_sequence_response(self.json_sequence_endpoint, i, sequence_start, sequence_range, expected_sequence_no=expected_seq_no)

    def test_increment_wraps_json_sequence(self):
        sequence_start = 1
        sequence_range = 1000000000

        test_start = 9999999998
        test_end = 1000000005

        for i in range(test_start, test_end):
            expected_seq_no = i % sequence_range
            self.mock_sequence_response(self.json_sequence_endpoint, i, sequence_start, sequence_range, expected_sequence_no=expected_seq_no)


server.py
"""Scalable service for generating sequences for SDX (backed by Postgres)."""
import logging
import os

from flask import abort, Flask, jsonify, request
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import exc, event, select
from sqlalchemy.exc import SQLAlchemyError
from structlog import wrap_logger
import psycopg2

import settings
from sequences import sequence, batch_sequence, image_sequence, json_sequence

__version__ = "2.6.4"

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = settings.DB_URL
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

logging.basicConfig(format=settings.LOGGING_FORMAT,
                    datefmt="%Y-%m-%dT%H:%M:%S",
                    level=settings.LOGGING_LEVEL)

logger = wrap_logger(logging.getLogger(__name__))

db = SQLAlchemy(app)
logger.info("START", version=__version__)

if os.getenv("CREATE_SEQUENCES", False):
    sequence.create(bind=db.engine)
    batch_sequence.create(bind=db.engine)
    image_sequence.create(bind=db.engine)
    json_sequence.create(bind=db.engine)


def sequence_values(seq, n=1):
    logger.debug("Obtaining next {0} of sequence".format(n))
    while n:
        try:
            result = db.engine.execute(seq.next_value())
            yield result.first()[0]
            n -= 1
        except (psycopg2.Error, SQLAlchemyError) as e:
            logger.error("Error executing sequence", exception=str(e))
            return


@app.route('/sequence', methods=['GET'])
def do_get_sequence():
    """Get the next sequence number. Starts at 1000 and increments to 9999."""
    sequence_start = 1000
    sequence_range = 9000
    try:
        n = int(request.args.get("n", 1))
    except (TypeError, ValueError):
        return abort(400)

    rv = {
        "sequence_list": [
            (i - 1) % sequence_range + sequence_start
            for i in sequence_values(sequence, n)
        ]
    }

    if n == 1:
        rv["sequence_no"] = rv["sequence_list"][0]

    return jsonify(rv)


@app.route('/batch-sequence', methods=['GET'])
def do_get_batch_sequence():
    """Get the next batch sequence number. Starts at 30000 and increments to 39999."""
    sequence_start = 30000
    sequence_range = 10000
    try:
        n = int(request.args.get("n", 1))
    except (TypeError, ValueError):
        return abort(400)

    rv = {
        "sequence_list": [
            (i - 1) % sequence_range + sequence_start
            for i in sequence_values(batch_sequence, n)
        ]
    }

    if n == 1:
        rv["sequence_no"] = rv["sequence_list"][0]

    return jsonify(rv)


@app.route('/image-sequence', methods=['GET'])
def do_get_image_sequence():
    """Get the next batch sequence number. Starts at 1 and increments to 999999999."""
    sequence_range = 1000000000
    try:
        n = int(request.args.get("n", 1))
    except (TypeError, ValueError):
        return abort(400)

    rv = {
        "sequence_list": [
            i % sequence_range
            for i in sequence_values(image_sequence, n)
        ]
    }

    if n == 1:
        rv["sequence_no"] = rv["sequence_list"][0]

    return jsonify(rv)


@app.route('/json-sequence', methods=['GET'])
def do_get_json_sequence():
    """Get the next sequence number for json files. Starts at 1 and increments to 999999999."""
    sequence_range = 1000000000
    try:
        n = int(request.args.get("n", 1))
    except (TypeError, ValueError):
        return abort(400)

    rv = {
        "sequence_list": [
            i % sequence_range
            for i in sequence_values(json_sequence, n)
        ]
    }

    if n == 1:
        rv["sequence_no"] = rv["sequence_list"][0]

    return jsonify(rv)


@app.route('/info', methods=['GET'])
@app.route('/healthcheck', methods=['GET'])
def healthcheck():
    try:
        conn = db.engine.connect()
        test_sql(conn)
    except SQLAlchemyError as e:
        logger.error("Failed to connect to database", exception=str(e))
        abort(500)
    return jsonify({'status': 'OK'})


@event.listens_for(db.engine, "engine_connect")
def test_connection(connection, branch):
    if branch:
        # "branch" refers to a sub-connection of a connection,
        # we don't want to bother testing on these.
        return

    save_should_close_with_result = connection.should_close_with_result
    connection.should_close_with_result = False
    try:
        logger.debug("Testing pooled connection")
        test_sql(connection)
        logger.debug("Connection successful")
    except exc.DBAPIError:
        try:
            logger.warning("Pooled connection failed - retrying")
            # try it again, this will recreate a stale or broke connection
            connection.scalar(select([1]))
            logger.debug("Connection successful")
        except exc.DBAPIError:
            # However if we get a database err again, forcibly close the connection
            logger.warning("Connection failed again - removing from pool")
            connection.close()
    finally:
        # restore "close with result"
        connection.should_close_with_result = save_should_close_with_result


def test_sql(connection):
    # Run a SELECT 1 to test the database connection
    logger.debug("Executing select 1")
    connection.scalar(select([1]))


if __name__ == '__main__':
    # Startup
    port = int(os.getenv("PORT"))
    app.run(debug=True, host='0.0.0.0', port=port)


settings.py
import logging
import os
from structlog import wrap_logger

LOGGING_LEVEL = logging.getLevelName(os.getenv('LOGGING_LEVEL', 'DEBUG'))
LOGGING_FORMAT = "%(asctime)s.%(msecs)06dZ|%(levelname)s: sdx-sequence: %(message)s"

logging.basicConfig(format=LOGGING_FORMAT,
                    datefmt="%Y-%m-%dT%H:%M:%S",
                    level=LOGGING_LEVEL)

logger = wrap_logger(
    logging.getLogger(__name__)
)


def _get_value(key):
    value = os.getenv(key)
    if not value:
        logger.error("No value set for {}".format(key))
        raise ValueError()
    else:
        return value


def build_db_url():
    db_host = _get_value("SDX_SEQUENCE_POSTGRES_HOST")
    db_port = _get_value('SDX_SEQUENCE_POSTGRES_PORT')
    db_name = _get_value('SDX_SEQUENCE_POSTGRES_NAME')
    db_user = _get_value('SDX_SEQUENCE_POSTGRES_USER')
    db_password = _get_value('SDX_SEQUENCE_POSTGRES_PASSWORD')
    db_url = 'postgres://{}:{}@{}:{}/{}'.format(db_user, db_password, db_host, db_port, db_name)
    return db_url


try:
    DB_URL = build_db_url()

except ValueError:
    logger.error("Unable to start service - DB connection details not set")


pgsequences.py
import os

import psycopg2
from psycopg2.pool import ThreadedConnectionPool


def get_dsn(settings=None):
    rv = {
        k: getattr(settings, v)
        for k, v in (
            ("host", "DB_HOST"), ("port", "DB_PORT"),
            ("dbname", "DB_NAME"), ("user", "DB_USER"), ("password", "DB_PASSWORD")
        )
    }
    return rv


class SequenceStore:

    class SQLSequence:
        seqs = {
            "sequence": slice(1000, 9999),
            "batch_sequence": slice(30000, 39999),
            "image_sequence": slice(1, 999999999),
            "json_sequence": slice(1, 999999999)
        }

        @staticmethod
        def sql():
            raise NotImplementedError

        def __init__(self, seq, criteria=None):
            self.seq = seq

        def run(self, con, cur=None, log=None):
            """
            Execute the SQL defined by this class.
            Returns the cursor for data extraction.

            """
            cur = cur or con.cursor()
            if self.seq in self.seqs:
                try:
                    cur.execute(self.sql(self.seq))
                except psycopg2.ProgrammingError:
                    con.rollback()
                else:
                    con.commit()
            return cur

    class Creation(SQLSequence):

        @staticmethod
        def check(seq):
            return (
                "SELECT 0 from pg_class where relname = '{seq}'"
            ).format(seq=seq)

        @staticmethod
        def sql(seq):
            criteria = SequenceStore.Creation.seqs.get(seq)
            return (
                "CREATE SEQUENCE {seq} MINVALUE {criteria.start} "
                "MAXVALUE {criteria.stop} CYCLE"
            ).format(seq=seq, criteria=criteria)

        def run(self, con, log=None):
            cur = con.cursor()
            cur.execute(self.check(self.seq))
            if not cur.fetchone():
                cur = super().run(con, cur)
            cur.close()

    class Query(SQLSequence):

        @staticmethod
        def sql(seq):
            return "SELECT nextval('{0}')".format(seq)

        def run(self, con, log=None):
            cur = super().run(con)
            try:
                row = cur.fetchone()
                return row[0] if row else None
            except psycopg2.ProgrammingError:
                return None
            finally:
                cur.close()


class ProcessSafePoolManager:
    """
    Connection pooling presents a challenge when gunicorn forks a worker
    after the pool is created. The result can be two worker processes claiming
    the same connection from the pool.

    See https://gist.github.com/jeorgen/4eea9b9211bafeb18ada for the basis of
    this solution.

    """

    @staticmethod
    def pool(*args, **kwargs):
        return ThreadedConnectionPool(*args, **kwargs)

    def __init__(self, **kwargs):
        self.pidLastSeen = os.getpid()
        self.kwargs = kwargs
        self._pool = None

    def getconn(self):
        pidNow = os.getpid()
        if self._pool is None or self._pool.closed or pidNow != self.pidLastSeen:
            minconn = self.kwargs.pop("minconn", 1)
            maxconn = self.kwargs.pop("maxconn", 16)
            self._pool = self.pool(minconn, maxconn, **self.kwargs)
            self.pidLastSeen = pidNow
        return self._pool.getconn()

    def putconn(self, conn):
        return self._pool.putconn(conn)

    def closeall(self):
        return self._pool.closeall()
